README.md
# README

This project implements a technique for Face Recognition in Keras based on [VGGFace2](http://www.robots.ox.ac.uk/~vgg/data/vgg_face2/). It recognizes the people that are referenced in the dataset (see the list [here](http://www.robots.ox.ac.uk/~vgg/data/vgg_face2/meta/identity_meta.csv)).

This [paper](https://arxiv.org/pdf/1710.08092.pdf) explains what is the dataset named `VGGFace2`.


## Quick Startup

### Download the network weights

You need to download the network `weights` [here](https://drive.google.com/file/d/1niCSaaqbXs6YP1B-DddRFQ7RAWsU8OW_/view?usp=sharing) and copy them in the `weights` folder.

### Download datasets

You can download a small dataset of images [here](https://drive.google.com/file/d/1m71-uiPA67NPPXqAfQwHVpdk79Ppj9vE/view?usp=sharing). Once, downloaded unzip the file in the `datasets` folder.

You also need to download the label file [here](https://drive.google.com/file/d/14-omxZDlz16zr1Xb15QkRfaPlu3NxSHA/view?usp=sharing). This file is used to convert the predictions of the `VGGFace` neural network to human readable information. Once, downloaded unzip the file in the `datasets` folder.

### Create a Virtual Environment

If you are on MacOS or Linux, type the following command to create a python virtual environment and install the required packages:

```bash
./configure.sh
```

Then, activate the virtual environment:

```bash
source .venv/bin/activate
```

***Nota***:
<br>This project requires `python 3`. You must install it first.


## Run

Execute the command:

```bash
python src/main.py -d 'dlib' -s './datasets/people/tim_cook.jpg'
```

You must get:

```bash
decoding ./datasets/people/tim_cook.jpg using ResNet50 and dlib face detector ...
Tim_Cook: 96.630%
Laurent_Ruquier: 0.258%
Mads_Gilbert: 0.069%
Rupert_Stadler: 0.061%
Jack_Layton: 0.048%
Done!
```

## How it works

The code is pretty easy to understand. The file `src/main.py` contains the class that performs the operations: decode face, predict and decode predictions.

```python
class Finder:
    def __init__(self):
        """Creates the VGGFace2 neural network, MTCNN and dlib face detectors."""
        self.vggface = ResNet50(n_classes=8631)
        self.vggface.load_weights(WEIGHTS)
        self.mtcnn = MTCNN()
        self.dlib = Dlib()

    def whois(self, image, detector, top=3):
      # First, extract the face from the passed-in image with 'mtcnn'
      # or 'dlib' face detector.
      if detector == 'mtcnn':
          face = _mtcnn_extract_face(self.mtcnn, image)
      elif detector == 'dlib':
          face = _dlib_extract_face(self.dlib, image)
      else:
          raise Exception('Sorry, only mtcnn and dlib decoders are allowed!')

      # Then, convert the face image from (224, 224, 3) to four dimensions,
      # and the amplitude from 0 - 255 to -127.5 - +127.5. Finally, perform
      # the predictions and convert these predictions to an human readable
      # information.
      x = face - 127.5
      image = np.expand_dims(x, axis=0)
      preds = self.vggface.predict(image)
      results = decode_predictions(preds, top)
      return results
```

The face is extracted from an image with `dlib` implemented in the file `src/decoders/dlib.py` or [MTCNN](https://github.com/ipazc/mtcnn).

The prediction is done through the `VGGFace` neural network implemented in the file `src/models/resnet50.py`.

Then, the predictions are decoded by the function `decode_predictions` implemented in the file `src/util/util.py`.


## License

MIT.


CHANGELOG.md
### HEAD

### 0.0.0 (June 12, 2020)

  * Initial build,
  * Release.


configure.sh
#!/bin/sh
#
# A bash script to create a virtual envrionment and install the python packages.
#
# copyright 2020 Artelsys <contact@artelsys.com> (http://www.artelsys.com/)
#
#
# Redistribution and use of this script, with or without modification, is
# permitted provided that the following conditions are met:
#
# 1. Redistributions of this script must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#
#  THIS SOFTWARE IS PROVIDED BY THE AUTHOR "AS IS" AND ANY EXPRESS OR IMPLIED
#  WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
#  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
#  EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
#  OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
#  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
#  OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
#  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

VENV=".venv"


# --- Private Functions --------------------------------------------------------

##
# Dumps the help message.
#
function help() {
  echo ''
    echo 'Usage: -v <virtual environment>'
    echo ''
    echo '-h  print the help function,'
    echo "-v  the name of the virtual environment. Default is ${VENV},"
    echo '-f  force to replace the existing virtual environment,'
    echo ''
}


# --- Main program -------------------------------------------------------------

venv=${VENV}

# Collect the passed in options:
while getopts "hv:f" opt; do
  case ${opt} in

    h ) help
        exit 1
      ;;

    v ) venv=$(echo ${OPTARG} | tr '[:upper:]' '[:lower:]')
        ;;

    f ) FORCE=true
        ;;

    : ) echo "Option -$OPTARG requires an argument" >&2
        help
        exit 1
        ;;

  esac
done


# Check if the virtual environment already exist?
if [[ -d ${venv} ]]
then
  if [[ ${FORCE} = true ]]
    then
      rm -rf ${venv}
    else
      echo "the virtual environment ${venv} already exists! Use the option -f to replace it."
      exit 1
  fi
fi


# Create the virtual environment
echo "Creating the virtual environment ${venv} ...:"
python3 -m venv ./${venv}
echo '  done.'

# Activate the environment
echo 'Activating the environment ...:'
source ./${venv}/bin/activate
echo '  done.'

# Update pip and setuptools
echo 'Updating pip and setuptools ...:'
pip install --upgrade pip setuptools
echo '  done.'
echo ''

# Install the packages required by Atom-IDE
echo 'Installing the packages required by Atom ...:'
source ./scripts/python_atom_packages_install.sh
echo '  done.'
echo ''

# Install the packages required by the project
echo 'Installing the packages required by the project ...:'
source ./scripts/python_project_packages_install.sh
echo '  done.'
echo ''

# reference your virtual environment to `Jupyter`
echo "Adding this virtual environment ${venv} to Jupyter kernels:"
python -m ipykernel install --user --name=${venv}
echo '  done.'

echo 'Done!'

# -- o ---


src/util/util.py
# ******************************************************************************
"""
A set of utility functions.


Private Functions:
    . none.


Public Functions:
    . decode_predictions        decodes the predictions returned by VGGFace2.


@namespace      _
@author         <author_name>
@since          0.0.0
@version        0.0.0
@licence        MIT. Copyright (c) 2020 Mobilabs <contact@mobilabs.fr>
"""
# ******************************************************************************
import numpy as np

LABEL_PATH = './datasets/labels_v1_8631.npy'


# -- Public Functions ----------------------------------------------------------

def decode_predictions(predictions, top=3):
    """Decodes the predictions returned by VGGFace2.

    ### Parameters:
        param1 (arr):       the resulting predictions.
        param2 (num):       the number of the most valuable predictions.

    ### Returns:
        (arr):              returns the matching people names,

    ### Raises:
        ValueError          if the predictions isn't a 2D array.
        ValueError          if the prediction doesn't contain 8631 values.
    """

    # predictions must be an array of array containing 8631 float numbers:
    # [[n0, n1, ... n8630]]
    if len(predictions.shape) == 2:
        if predictions.shape[1] == 8631:
            LABELS = np.load(LABEL_PATH)
        else:
            raise ValueError('decode_predictions expects an array with 8631 values!')

    else:
        raise ValueError('decode_predictions expects a 2D array!')

    # Extract the top predictions:
    results = []
    for pred in predictions:
        top_indices = pred.argsort()[-top:][::-1]
        for i in top_indices:
            # results.append([str(LABELS[i].encode('utf8')), pred[i]])
            results.append([str(LABELS[i]), pred[i]])

    return results

# -- o ----


src/decoders/dlib.py
# ******************************************************************************
"""
Detects or extracts a face(s) from an image with the dlib decoder.


Private Functions:
    . none,


Public Class:
    .  Dlib                     a class to detect or extract a face from an image,


Private Methods:
    . __detect_face             detects the face(s) on the passed-in image,


Public Methods:
    . get_face                  extracts the face from the passed-in image,
    . get_image_with_faces      detects and highlights the face(s) on the image,


@namespace      _
@author         <author_name>
@since          0.0.0
@version        0.0.0
@licence        MIT. Copyright (c) 2020 Mobilabs <contact@mobilabs.fr>
"""
# ******************************************************************************
import cv2
import dlib

RECT_COLOR = (0, 255, 255)
RECT_THICKNESS = 2


# -- Public --------------------------------------------------------------------

class Dlib:
    """A class to detect or extract a face from an image with dlib.

    ### Attributes:
        classifier (obj):       the dlib decoder object.

    ### Methods:
        get_face(image):
            extracts the face from the passed-in image.

        get_image_with_faces(image):
            detects and highlights the face(s) on the passed-in image.

    ### Raises:
        none
    """

    def __init__(self):
        """Creates the dlib decoder."""
        self.detector = dlib.get_frontal_face_detector()

    def __detect_face(self, img):
        """Detects the face(s) on the passed-in image.

        ### Parameters:
            param1 (str):   the path of the input image.

        ### Returns:
            (arr):          returns the coordinates of the detected face(s).

        ### Raises:
            -
        """
        gray = cv2.cvtColor(img.copy(), cv2.COLOR_BGR2GRAY)
        return self.detector(gray, 1)

    def get_face(self, image):
        """Extracts the face from the passed-in image.

        ### Parameters:
            param1 (arr):   the input image.

        ### Returns:
            (arr):          returns the extracted face.

        ### Raises:
            -
        """
        face = self.__detect_face(image)[0]
        x1, y1, x2, y2, _, _ = face.left(), face.top(), \
            face.right() + 1, face.bottom() + 1, face.width(), face.height()
        return image[y1:y1 + (y2 - y1), x1:x1 + (x2 - x1), :]

    def get_image_with_faces(self, image):
        """Detects and highlights the face(s) on the passed-in image.

        ### Parameters:
            param1 (arr):   the input image.

        ### Returns:
            (arr):          returns the input image with highlighted face(s).

        ### Raises:
            -
        """
        img = image.copy()
        faces = self.__detect_face(img)

        if len(faces) > 0:
            for i, d in enumerate(faces):
                x1, y1, x2, y2, _, _ = d.left(), d.top(), \
                    d.right() + 1, d.bottom() + 1, d.width(), d.height()
                cv2.rectangle(img, (x1, y1), (x2, y2), RECT_COLOR, RECT_THICKNESS)
        return img

# -- o ---


src/models/resnet50.py
# ******************************************************************************
"""
Creates and Returns ResNet50 Convolutional Neural Network.
(referenced on https://arxiv.org/pdf/1512.03385.pdf)

This implementation is derived from the official keras implementation that
can be found here:
    . https://github.com/keras-team/keras-applications/blob/master/
      keras_applications/resnet50.py


Private Functions:
    . _conv_block                   a block that has a conv layer at shortcut,
    . _dw_conv_block                a block that has no conv layer at shortcut,


Public Functions:
    . ResNet50                      returns ResNet50 CNN,


@namespace      -
@author         Mobilabs
@since          0.0.0
@version        0.0.0
@licence        MIT. Copyright (c) 2020 Mobilabs <contact@mobilabs.fr>
"""
# ******************************************************************************
import keras
from keras.layers import Activation, Dense, Input
from keras.layers import Conv2D, ZeroPadding2D
from keras.layers import MaxPooling2D, GlobalAveragePooling2D
from keras.layers import BatchNormalization


# -- Private Functions ---------------------------------------------------------


def _conv_block(input_tensor, kernel_size, filters, stage, block, strides=(2, 2)):
    """The conv block is the block that has a conv layer at shortcut.

    ### Parameters:
        param1 (obj):       intermediate input tensor.
        param2 (num):       the size of the kernel.
        param3 (arr):       the size of the filters.
        param4 (num):       the identity block number.
        param5 (num):       the block number inside the identity block.
        param5 (tuple):     the size of the stride.

    ### Returns:
        (obj):              returns the intermediate output tensor.

    ### Raises:
        none
    """
    filters1, filters2, filters3 = filters
    name = 'conv' + str(stage) + '_' + str(block)

    name1 = name + '_1x1_reduce'
    x = Conv2D(filters1, (1, 1), strides=strides,
               kernel_initializer='he_normal',
               use_bias=False,
               padding='same',
               name=name1)(input_tensor)
    x = BatchNormalization(name=name1 + '_bn')(x)
    x = Activation('relu', name=name1 + '_relu')(x)

    name2 = name + '_3x3'
    x = Conv2D(filters2, kernel_size,
               kernel_initializer='he_normal',
               use_bias=False,
               padding='same',
               name=name2)(x)
    x = BatchNormalization(name=name2 + '_bn')(x)
    x = Activation('relu', name=name2 + '_relu')(x)

    name3 = name + '_1x1_increase'
    x = Conv2D(filters3, (1, 1),
               kernel_initializer='he_normal',
               use_bias=False,
               padding='same',
               name=name3)(x)
    x = BatchNormalization(name=name3 + '_bn')(x)

    name4 = name + '_1x1_proj'
    shortcut = Conv2D(filters3, (1, 1), strides=strides,
                      kernel_initializer='he_normal',
                      use_bias=False,
                      padding='same',
                      name=name4)(input_tensor)
    shortcut = BatchNormalization(name=name4 + '_bn')(shortcut)

    x = keras.layers.add([x, shortcut], name=name + '_add')
    x = Activation('relu', name=name + '_add_relu')(x)
    return x


def _identity_block(input_tensor, kernel_size, filters, stage, block):
    """The identity block is the block that has no conv layer at shortcut.

    ### Parameters:
        param1 (obj):       intermediate input tensor.
        param2 (num):       the size of the kernel.
        param3 (arr):       the size of the filters.
        param4 (num):       the identity block number.
        param5 (num):       the block number inside the identity block.

    ### Returns:
        (obj):              returns the intermediate output tensor.

    ### Raises:
        none
    """
    filters1, filters2, filters3 = filters
    name = 'conv' + str(stage) + '_' + str(block)

    name1 = name + '_1x1_reduce'
    x = Conv2D(filters1, (1, 1),
               kernel_initializer='he_normal',
               use_bias=False,
               padding='same',
               name=name1)(input_tensor)
    x = BatchNormalization(name=name1 + '_bn')(x)
    x = Activation('relu', name=name1 + '_relu')(x)

    name2 = name + '_3x3'
    x = Conv2D(filters2, kernel_size,
               kernel_initializer='he_normal',
               use_bias=False,
               padding='same',
               name=name2)(x)
    x = BatchNormalization(name=name2 + '_bn')(x)
    x = Activation('relu', name=name2 + '_relu')(x)

    name3 = name + '_1x1_increase'
    x = Conv2D(filters3, (1, 1),
               kernel_initializer='he_normal',
               use_bias=False,
               padding='same',
               name=name3)(x)
    x = BatchNormalization(name=name3 + '_bn')(x)

    x = keras.layers.add([x, input_tensor], name=name + '_add')
    x = Activation('relu', name=name + '_add_relu')(x)
    return x


# -- Public Functions ----------------------------------------------------------

def ResNet50(input_shape=(224, 224, 3), n_classes=1000):
    """Creates the ResNet50 CNN.

    ### Parameters:
        param1 (tuple):     the shape of the input image.
        param2 (num):       the number of classes.

    ### Returns:
        (obj):              returns the Resnet50 CNN model.

    ### Raises:
        none
    """
    # Create the Tensor
    input = Input(shape=input_shape)

    # inputs are of size 224 x 224 x 3
    x = Conv2D(64, (7, 7), strides=(2, 2),
               kernel_initializer='he_normal',
               use_bias=False,
               padding='valid',
               name='conv1_7x7_s2')(input)
    x = BatchNormalization(name='conv1_7x7_s2_bn')(x)
    x = Activation('relu', name='conv1_7x7_s2_relu')(x)
    x = ZeroPadding2D(padding=(1, 1), name='conv1_7x7_s2_pad_1x1')(x)
    x = MaxPooling2D(pool_size=(3, 3), strides=(2, 2), name='conv1_7x7_s2_pool_3x3')(x)

    x = _conv_block(x, 3, [64, 64, 256], stage=2, block=1, strides=(1, 1))
    x = _identity_block(x, 3, [64, 64, 256], stage=2, block=2)
    x = _identity_block(x, 3, [64, 64, 256], stage=2, block=3)

    x = _conv_block(x, 3, [128, 128, 512], stage=3, block=1)
    x = _identity_block(x, 3, [128, 128, 512], stage=3, block=2)
    x = _identity_block(x, 3, [128, 128, 512], stage=3, block=3)
    x = _identity_block(x, 3, [128, 128, 512], stage=3, block=4)

    x = _conv_block(x, 3, [256, 256, 1024], stage=4, block=1)
    x = _identity_block(x, 3, [256, 256, 1024], stage=4, block=2)
    x = _identity_block(x, 3, [256, 256, 1024], stage=4, block=3)
    x = _identity_block(x, 3, [256, 256, 1024], stage=4, block=4)
    x = _identity_block(x, 3, [256, 256, 1024], stage=4, block=5)
    x = _identity_block(x, 3, [256, 256, 1024], stage=4, block=6)

    x = _conv_block(x, 3, [512, 512, 2048], stage=5, block=1)
    x = _identity_block(x, 3, [512, 512, 2048], stage=5, block=2)
    x = _identity_block(x, 3, [512, 512, 2048], stage=5, block=3)

    # Output layer
    x = GlobalAveragePooling2D()(x)
    x = Dense(n_classes, name='predictions')(x)
    x = Activation('softmax', name='predictions_softmax')(x)

    model = keras.models.Model(input, x, name='ResNet50')
    return model


if __name__ == '__main__':
    model = ResNet50()
    model.summary()

    keras.utils.plot_model(model,
                           to_file='./diagrams/ResNet50.png',
                           show_shapes=True,
                           show_layer_names=True)

# -- o ---


src/main.py
# ******************************************************************************
"""
Finds people on the passed-in image.

Implements VGGFace2 on keras for performing people face recognition as described
here: http://www.robots.ox.ac.uk/~vgg/data/vgg_face2/.


Private Functions:
    . _parse                    parses the script arguments.
    . _mtcnn_extract_face       extracts a face using the MTCNN face detector.
    . _dlib_extract_face        extracts a face using the dlib face detector.


Public Class:
    .  Finder                   a class to find people on an image.


Public Methods:
    . whois                     finds who is the person on the image.


@namespace      _
@author         <author_name>
@since          0.0.0
@version        0.0.0
@licence        MIT. Copyright (c) 2020 Mobilabs <contact@mobilabs.fr>
"""
# ******************************************************************************
import argparse
from decoders.dlib import Dlib
from models.resnet50 import ResNet50
from util.util import decode_predictions

import numpy as np
import cv2
from PIL import Image
from matplotlib import pyplot
from mtcnn.mtcnn import MTCNN

WEIGHTS = './weights/vggface_tf_resnet50_v1_8631.h5'


# -- Private Functions ---------------------------------------------------------

def _parse():
    """Parses the script arguments.

    ### Parameters:
        param1 ():          none,

    ### Returns:
        (str):              returns the option values,

    ### Raises:
        none
    """
    parser = argparse.ArgumentParser()
    parser.add_argument('-s', '--source', type=str, help='path of the input image')
    parser.add_argument('-d', '--decoder', type=str, help='the face decoder mtcnn or dlib')
    args = parser.parse_args()

    if args.source is None or args.decoder is None:
        print('You must provide the path of the image to analyze and the decoder!')
        parser.print_help()
        exit()
    else:
        return args.source, args.decoder


def _mtcnn_extract_face(detector, image, required_size=(224, 224)):
    """Extracts a face from an image using the MTCNN face detector.

    ### Parameters:
        param1 (obj):       the MTCNN detector object.
        param2 (arr):       the input image.
        param3 (tuple):     the face size.

    ### Returns:
        (arr):              returns the detected face.

    ### Raises:
        none
    """
    # detect faces in the image and extract the bounding box
    # from the first face
    results = detector.detect_faces(image)
    x1, y1, width, height = results[0]['box']
    x2, y2 = x1 + width, y1 + height
    face = image[y1:y2, x1:x2]

    # resize the extracted face to the size required by
    # the VGGFace CNN
    image = Image.fromarray(face)
    image = image.resize(required_size)
    face_array = np.asarray(image)
    return face_array


def _dlib_extract_face(detector, image, required_size=(224, 224)):
    """Extracts a face from an image using the dlib face detector.

    ### Parameters:
        param1 (obj):       the Dlib detector object.
        param2 (arr):       the input image.
        param3 (tuple):     the face size.

    ### Returns:
        (arr):              returns the detected face.

    ### Raises:
        none
    """
    face = detector.get_face(image)
    face = cv2.resize(face, required_size, interpolation=cv2.INTER_LINEAR)
    return face


# -- Public --------------------------------------------------------------------

class Finder:
    """Class for a Face Recognition.

    ### Attributes:
        vggface (obj):      VGGFace2 Neural network.
        mtcnn (obj):        MTCNN face detector object.
        dlib (obj):         Dlib face detector object.

    ### Methods:
        whois(image, detector):
            finds who is the person on the image.

    ### Raises:
        Exception           if the detector isn't supported.
        """

    def __init__(self):
        """Creates the VGGFace2 neural network, MTCNN and dlib face detectors."""
        self.vggface = ResNet50(n_classes=8631)
        self.vggface.load_weights(WEIGHTS)
        self.mtcnn = MTCNN()
        self.dlib = Dlib()

    def whois(self, image, detector, top=3):
        """Finds who is the person on the image.

        ### Parameters:
            param1 (arr):   the input image.
            param2 (obj):   the face detector to use.
            param3 (num):   the number of the most valuable predictions.

        ### Returns:
            (arr):          returns the most valuable predictions.

        ### Raises:
            none
        """

        # First, extract the face from the passed-in image with 'mtcnn'
        # or 'dlib' face detector.
        if detector == 'mtcnn':
            face = _mtcnn_extract_face(self.mtcnn, image)
        elif detector == 'dlib':
            face = _dlib_extract_face(self.dlib, image)
        else:
            raise Exception('Sorry, only mtcnn and dlib decoders are allowed!')

        # Then, convert the face image from (224, 224, 3) to four dimensions,
        # and the amplitude from 0 - 255 to -127.5 - +127.5. Finally, perform
        # the predictions and convert these predictions to an human readable
        # information.
        x = face - 127.5
        image = np.expand_dims(x, axis=0)
        preds = self.vggface.predict(image)
        results = decode_predictions(preds, top)
        return results


if __name__ == '__main__':
    source, detector = _parse()
    finder = Finder()

    if detector == 'mtcnn':
        print('decoding ' + source + ' using ResNet50 and mtcnn face detector ...')
        image = pyplot.imread(source)
        results = finder.whois(image, 'mtcnn', top=5)

    elif detector == 'dlib':
        print('decoding ' + source + ' using ResNet50 and dlib face detector ...')
        image = cv2.imread(source)
        results = finder.whois(image, 'dlib', top=5)

    else:
        print('The face detector must be mtcnn or dlib')
        exit()

    # display the most likely results
    for result in results:
        print('%s: %.3f%%' % (result[0].lstrip(), result[1] * 100))
    print('Done!')

# -- o ----
